core/src/layers/logging.rs

```diff
@@ -82,14 +80,136 @@ use crate::*;
+        Self {
+            notify: self.notify.clone(),
+        }
+    }
+}
+
+impl Default for LoggingLayer {
+    fn default() -> Self {
+        Self {
+            notify: Arc::new(DefaultLoggingInterceptor::default()),
+        }
+    }
+}
+
+impl LoggingLayer {
+    /// Create the layer with specific logging interceptor.
+    ///
+    /// ```no_run
+    /// use crate::layers::LoggingInterceptor;
+    /// use crate::layers::LoggingLayer;
+    /// use crate::services;
+    /// use crate::Error;
+    /// use crate::Operator;
+    /// use crate::Scheme;
+    ///
+    /// #[derive(Debug, Clone)]
+    /// struct MyLoggingInterceptor;
+    ///
+    /// impl LoggingInterceptor for MyLoggingInterceptor {
+    ///     fn log(
+    ///         &self,
+    ///         scheme: Scheme,
+    ///         operation: &'static str,
+    ///         path: &str,
+    ///         message: &str,
+    ///         err: Option<&Error>,
+    ///     ) {
+    ///         // log something
+    ///     }
+    /// }
+    ///
+    /// let _ = Operator::new(services::Memory::default())
+    ///     .expect("must init")
+    ///     .layer(LoggingLayer::new(MyLoggingInterceptor))
+    ///     .finish();
+    /// ```
+    ///
+    /// This allows customize the log format.
+    pub fn new<I: LoggingInterceptor>(notify: I) -> LoggingLayer<I> {
+        LoggingLayer {
+            notify: Arc::new(notify),
+        }
+    }
+}
+
+impl<A: Access, I: LoggingInterceptor> Layer<A> for LoggingLayer<I> {
+    type LayeredAccess = LoggingAccessor<A, I>;
+
+    fn layer(&self, inner: A) -> Self::LayeredAccess {
+        let meta = inner.info();
+        LoggingAccessor {
+            inner,
+
+            ctx: LoggingContext {
+                scheme: meta.scheme(),
+                notify: self.notify.clone(),
+            },
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct LoggingContext<I> {
+    scheme: Scheme,
+    notify: Arc<I>,
+}
+
+impl<I> Clone for LoggingContext<I> {
+    fn clone(&self) -> Self {
+        Self {
+            scheme: self.scheme,
+            notify: self.notify.clone(),
+        }
+    }
+}
+
+impl<I: LoggingInterceptor> LoggingContext<I> {
+    fn log(&self, operation: &'static str, path: &str, message: &str, err: Option<&Error>) {
+        self.notify.log(self.scheme, operation, path, message, err)
+    }
+}
+
+/// LoggingInterceptor is used to intercept the log.
+pub trait LoggingInterceptor: Debug + Send + Sync + 'static {
+    /// Everytime there is a log, this function will be called.
+    ///
+    /// # Inputs
+    ///
+    /// - scheme: The service generates the log.
+    /// - operation: The operation to log.
+    /// - path: The path argument to the operator, maybe empty if the
+    ///         current operation doesn't have a path argument.
+    /// - message: The log message.
+    /// - err: The error to log.
+    fn log(
+        &self,
+        scheme: Scheme,
+        operation: &'static str,
+        path: &str,
```

Comment:
How about add `context: &str` instead of just `path`? We can add `read`/`write`/`from`/`to` in context at the same time.

Reply:
Let me have a try.